File: practice/src/view_video.cpp
#include <opencv4/opencv2/highgui/highgui.hpp>
#include <opencv4/opencv2/imgproc/imgproc.hpp>
#include <string>
#include <iostream>

cv::VideoCapture g_cap;
int g_slider_position;
int g_run = -1;
bool g_pauseit = false;

void onTrackbarSlide(int pos, void*){
    g_cap.set(cv::CAP_PROP_POS_FRAMES, pos);
    if(g_pauseit) g_run = 1;
    g_pauseit = true;
}

int main(int argc, char** argv){
    std::string WINDOW_NAME = "video";
    cv::namedWindow(WINDOW_NAME);
    try{
        g_cap.open(std::string(argv[1]));
    }
    catch(...){
        std::cout<<"\nCannot open file";
        exit(0);
    }
    int frames = (int) g_cap.get(cv::CAP_PROP_FRAME_COUNT);
    int fw = (int) g_cap.get(cv::CAP_PROP_FRAME_WIDTH);
    int fh = (int) g_cap.get(cv::CAP_PROP_FRAME_HEIGHT);
    std::cout<<"\nVideo has "<<frames<<" frames with resolution: "<<fh<<"x"<<fw;
    cv::Mat frame;
    cv::createTrackbar("Position", WINDOW_NAME, &g_slider_position, frames, onTrackbarSlide);
    while(true){
        if(g_run!=0){
            g_cap>>frame;
            if(frame.empty()) break;
            int current_pos = (int) g_cap.get(cv::CAP_PROP_POS_FRAMES);
            g_pauseit = false;
            cv::setTrackbarPos("Position", WINDOW_NAME, current_pos);
            cv::imshow(WINDOW_NAME, frame);
            g_run--;
        }
        char c = cv::waitKey(10);
        if(c == 's' || c == 'S'){
            g_run = 1;
        }
        if(c == 'r' || c == 'R'){
            g_run=-1;
        }
        if(c == 27) break;
    }
    std::cout<<"\nVideo complete or 'quit' key pressed!\n";
    cv::destroyWindow(WINDOW_NAME);
    return 0;
}

File: practice/scripts/view_video.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# This program plays back a video file
# with seek and pause features
# 
# author: Anshay
# Date: 06/01/2021


# In[ ]:


import numpy as np
import cv2
import sys


# In[ ]:


## Global Variables
g_cap = cv2.VideoCapture()
g_play = True
print(g_play)


# In[ ]:


def onTrackbarSlide(pos):
    global g_play
    g_cap.set(cv2.CAP_PROP_POS_FRAMES, pos)
    g_play = False


# In[ ]:


def view_video(filepath):
    global g_play
    g_cap.open(filepath)
    trackbar_name = "Slider"
    window_name = "Video Play"
    cv2.namedWindow(window_name)
    max_frames = int(g_cap.get(cv2.CAP_PROP_FRAME_COUNT))
    cv2.createTrackbar(trackbar_name, window_name, 0, max_frames, onTrackbarSlide)
    ret, frame = g_cap.read()
    while(ret):
        if(g_play):
            frame = cv2.resize(frame, (640, 480))
            cv2.imshow(window_name, frame)
            ret, frame = g_cap.read()
        key = cv2.waitKey(30)
        if(key==27):
            break
        if(key==ord('p') or key==ord('P')):
            g_play = not g_play
    g_cap.release()
    print("\nVideo finished or error occured")


# In[ ]:


if __name__== "__main__":
    if(len(sys.argv) != 2):
        print("\nCheck correct usage: 'view_video <filepath>' ")
        sys.exit()
    filename = sys.argv[1]
    view_video(filename)
    cv2.destroyAllWindows()


# In[ ]:




